tpm_futurepcr/binary_reader.py

```diff
@@ -1,53 +1,46 @@
+import io
 import struct
+from pathlib import Path
 
-class BinaryReader():
-    def __init__(self, fh):
-        self.fh = fh
-        self._N_size = struct.calcsize("@N")
-        self._P_size = struct.calcsize("@P")
+_NSIZE = struct.calcsize("@N")
+_PSIZE = struct.calcsize("@P")
 
-    def seek(self, pos, whence=0):
-        return self.fh.seek(pos, whence)
-
-    def read(self, length):
-        buf = self.fh.read(length)
-        if len(buf) < length:
-            if len(buf) == 0:
-                raise EOFError("Hit EOF after 0/%d bytes" % length)
-            else:
-                raise IOError("Hit EOF after %d/%d bytes" % (len(buf), length))
-        return buf
 
-    def _read_fmt(self, length, fmt):
-        buf = self.fh.read(length)
-        if len(buf) < length:
-            if len(buf) == 0:
-                raise EOFError("Hit EOF after 0/%d bytes" % length)
-            else:
-                raise IOError("Hit EOF after %d/%d bytes" % (len(buf), length))
-        data, = struct.unpack(fmt, buf)
-        return data
+class ReadFormats:
+    U8 = "B"
+    U16 = "H"
+    U32 = "L"
+    U64 = "Q"
+    PTR = "Q" if _PSIZE == 8 else "L"
+    SIZE = "Q" if _NSIZE == 8 else "L"
 
-    def read_u8(self):
-        return self._read_fmt(1, ">B")
 
-    def read_u16_le(self):
-        return self._read_fmt(2, "<H")
+class BinaryReader:
+    def __init__(self, fh_buf: bytes | Path):
+        self.fh = open(fh_buf, "rb") if isinstance(fh_buf, Path) else io.BytesIO(fh_buf)
 
-    def read_u32_le(self):
-        return self._read_fmt(4, "<L")
+    def seek(self, pos: int, whence: int = 0):
+        return self.fh.seek(pos, whence)
 
-    def read_u64_le(self):
-        return self._read_fmt(8, "<Q")
+    def read(self, fmt_len: str | int) -> bytes | int | tuple | None:
```

Comment:
I kind of really don't want to accept this conversion in particular – it doesn't feel like it would improve either readability or usability (or even type-checking if it replaces everything with one super-function that can return any type).

Reply:
Hmm, I don't think they're needed _currently_ – I think originally I added them because it wasn't clear to me whether the endianness varies by architecture (like e.g. the size of `image_length` in parse_efi_bsa_event varies by architecture) or whether it's always LE.

If it turns out that the endianness is system-dependent – yeah, a CLI flag to force a specific endianness may be useful for tests (and/or the separate "log pretty-print" tool in the future). I think it could be passed in as a parameter to the BinaryReader constructor. Similarly, an option to force a specific size of UINTN (32 or 64 bits) may be useful in case someone has a 32-bit UEFI or wants to run tests on a 32-bit host...

Reply:
I have given a fast look to the TPM v1/2 specification, but I could not find anything related to the format and endianess of those logs. Seems that for TPM v2 they are little endian, but in other sources they say the structures are stored on native endianess. I'd suggest to put this on hold until we can get some logs from these other architectures, and then see what needs to be changed and how, and proceed. What do you think?

Reply:
@grawity, did you see my last message?